lambda-buffers-compiler/test/Test/DeriveCheck.hs

```diff
@@ -0,0 +1,627 @@
+test =
+  testGroup
+    "Derive Tests"
+    [ test1
+    , test2
+    , test3
+    , test4
+    , test5
+    , test6
+    , test7
+    , test8
+    , test9
+    , test10
+    , test11
+    ]
+  where
+    test1, test2, test3, test4, test5, test6, test7, test8, test9, test10, test11 :: TestTree
+
+    test1 =
+      testCase "1: Basic simple module" $
+        assertPrism "FAIL!" _Right (runTest "A" moduleA'1)
+
+    test2 =
+      testCase "2: Module w/ imports" $
+        assertPrism "FAIL!" _Right (runTest "B" moduleB'1)
+
+    test3 =
+      testCase "3: Imports w/ tyvars" $
+        assertPrism "FAIL!" _Right (runTest "B" moduleB'2)
+
+    test4 =
+      testCase "4: Circular constraints" $
+        assertPrism
+          "FAIL!"
+          (_Left . _Ctor @"BadInstance" . _Ctor @"TyConInContext")
+          (runTest "B" moduleB'3)
+
+    test5 =
+      testCase "5: Overlapping instances" $
+        assertPrism
+          "FAIL!"
+          (_Left . _Ctor @"BadInstance" . _Ctor @"OverlapDetected")
+          (runTest "B" moduleB'4)
+
+    test6 =
+      testCase "6: Missing Int instance" $
+        assertCtor @"CouldntSolveConstraints"
+          "FAIL!"
+          (runTest "B" moduleB'5)
+
+    test7 =
+      testCase "7: Constr in instance constraint" $
+        assertPrism
+          "FAIL!"
+          (_Left . _Ctor @"BadInstance" . _Ctor @"TyConInContext")
+          (runTest "B" moduleB'6)
+
+    test8 =
+      testCase "8: Multiple TyVar data type" $
+        assertPrism
+          "FAIL!"
+          _Right
+          (runTest "C" moduleC'1)
+
+    test9 =
+      testCase "9: Phantom tyvar arg" $
+        assertPrism
+          "FAIL"
+          _Right
+          (runTest "D" moduleD'1)
+
+    test10 =
+      testCase "10: Phantom (concrete) type arg" $
+        assertPrism
+          "FAIL!"
+          _Right
+          (runTest "D" moduleD'2)
+
+    test11 =
+      testCase "11: Missing type in phantom" $
+        assertCtor @"CouldntSolveConstraints"
+          "FAIL!"
+          (runTest "D" moduleD'3)
+
+-- We should be able to avoid tricky superclass examples here
+-- b/c we have them in the test suite for `solve`
+
+pattern (:@) :: Pat -> Pat -> Pat
+pattern (:@) p1 p2 = AppP p1 p2
+
+pattern (:@@) :: Exp -> Exp -> Exp
+pattern (:@@) p1 p2 = AppE p1 p2
+
+pattern LocalRefP :: Text -> Pat
+pattern LocalRefP nm = RefP NilP (LitP (Name nm))
+
+pattern ForeignRefP :: [Text] -> Text -> Pat
+pattern ForeignRefP mn nm = RefP (LitP (ModuleName mn)) (LitP (Name nm))
+
+pattern ForeignRefE :: [Text] -> Text -> Exp
+pattern ForeignRefE mn nm = RefE (LitE (ModuleName mn)) (LitE (Name nm))
+
+pattern OpaqueDecNoVars :: Text -> Exp
+pattern OpaqueDecNoVars nm = DecE (LitE (Name nm)) NilE (LitE Opaque)
+
+moduleName1 :: Text -> PC.ModuleName
+moduleName1 nm = PC.ModuleName [PC.ModuleNamePart nm si] si
+  where
+    si :: PC.SourceInfo
+    si = PC.SourceInfo "" pos pos
+    pos :: PC.SourcePosition
+    pos = PC.SourcePosition 0 0
```

Comment:
Honestly this was just something stupid to provide a `ModuleName` to `runDeriveCheck` (which is intended to work on key-value pairs in a `Map ModuleName ModuleBuilder`). Because we're only looking at individual modules (w/ imports constructed via artifice) I think I can just use an empty module. 

Unless I'm misunderstanding your points? I'm going to move this into a `where` clause inside `runTest` and just use an empty `ModuleName`